feat(matching-service): implement basic matching algorithm and endpoint

ruiming97 opened the pull request:
### Description:
This PR introduces the foundational matching logic for the Matching Service. It sets up the API endpoint and implements an in-memory, exact-match algorithm to pair users based on their selected criteria.

#### Changes Included:
API Endpoint: Created a POST /api/matches endpoint to handle new match requests from users.

In-Memory Queue: Implemented an in-memory Map to act as a real-time waiting queue for users.
_Note that this is temporary, and will be converted into Redish Cache in later sprint_

Matching Logic: Added the core algorithm to create a key from user criteria (difficulty, topics) and find an **exact** match (all selected criteria are matched) in the queue.

Code Structure: Refactored the service into a standard controller/service/model architecture for better organization and scalability.

Error Handling: Integrated the shared errorHandler for consistent error responses.

#### How to Test:
1. Run the stack with docker compose up --build.

2. Send a POST request to http://localhost:8081/api/matches with the following body in JSON:
```
{
  "userId": "user-1",
  "criteria": {
    "difficulties": ["Medium"],
    "topics": ["Arrays", "Hash Table"]
  }
}
```
3. Verify that the server responds with a 202 Accepted status and {"status": "waiting"}.

4. Send a second, identical request with a different userId (e.g., "user-2").

5. Verify that the server responds with a 200 OK status and a body containing the matched partner's ID (e.g., {"status": "matched", "partnerId": "user-1", ...}).

### Relevant Jira Stories / Subtasks
closes PP-32
relates to PP-64
closes PP-65
closes PP-67

ruiming97 committed:
feat(matching-service): add basic matching criteria endpoint and logic

ChinZJ commented on matching-service/src/constants/env.ts:
```diff
@@ -0,0 +1,32 @@
+
+// Solution adapted from:
+// https://stackoverflow.com/questions/64383909/dirname-is-not-defined-error-in-node-js-14-version
+const filename = fileURLToPath(import.meta.url);
```
Im not sure if we may need to further split our packages to be service dependent instead as I had the following issue:

Jest runs in `cjs` while our runtime is in `esnext`. Thus there will be an issue compiling this code as well. The fix I did was to switch to `process.cwd()` and backtrack until the root folder is reached.

For now it works because the root folder is the name of the repo but definitely not ideal

ChinZJ commented on matching-service/src/services/matchingService.ts:
```diff
@@ -0,0 +1,31 @@
+import type { MatchCriteria } from '../models/matchModel';
+
+const waitingQueue: Map<string, string[]> = new Map<string, string[]>;
```
```suggestion
const waitingQueue: Map<string, string[]> = new Map<string, string[]>();
```

ChinZJ commented on matching-service/src/services/matchingService.ts:
```diff
@@ -0,0 +1,31 @@
+}
+// example criteria key: Easy,Medium-Arrays,HashMap
+
+export const findOrQueueUser = (userId: string, criteria: MatchCriteria) => {
```
Seems susceptible to race conditions.

Also needs mechanism to remove stale users

ruiming97 replied:
Yes will handle them in next PR

ChinZJ reviewed (commented):
Unsure if meant to be complete (see comments). Else LGTM!

ruiming97 committed:
feat(matching-service): add WebSocket for real-time match notifications

Implements WebSocket capabilities in the matching service to enable real-time notifications for queued users.

- Configures a WebSocket server alongside the existing HTTP server.
- Establishes a mechanism to track active WebSocket connections by user ID.
- Modifies matching logic to send a `match_found` event over WebSocket to waiting users upon successful pairing.
- Includes WebSocket connection registration and cleanup handlers.
- Updates `findOrQueueUser` to leverage WebSocket for proactive notifications.

ruiming97 committed:
Update matching-service/src/services/matchingService.ts

Co-authored-by: ChinZJ <122326502+ChinZJ@users.noreply.github.com>

verakohh reviewed (approved):
lgtm

ruiming97 committed:
Merge branch 'develop' into matching-service/basic-matching-criteria

# Conflicts:
#	package-lock.json
#	package.json

ruiming97 committed:
Fix merge conflict
